src/main/java/nextstep/security/VirtualFilterChain.java

```diff
@@ -0,0 +1,31 @@
+package nextstep.security;
+
+import jakarta.servlet.Filter;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.ServletResponse;
+
+import java.io.IOException;
+import java.util.List;
+
+public class VirtualFilterChain implements FilterChain {
+
+    private final List<Filter> filters;
+    private final FilterChain originalChain;
+    private int currentPosition = 0;
+
+    public VirtualFilterChain(List<Filter> filters, FilterChain originalChain) {
+        this.filters = filters;
+        this.originalChain = originalChain;
+    }
+
+    public void doFilter(ServletRequest request, ServletResponse response) throws IOException, ServletException {
+        if (this.currentPosition == filters.size()) {
+            this.originalChain.doFilter(request, response);
+        } else {
+            Filter nextFilter = filters.get(currentPosition++);
+            nextFilter.doFilter(request, response, this);
+        }
```

Comment:
```suggestion
        if (this.currentPosition == filters.size()) {
            this.originalChain.doFilter(request, response);
            return;
        }
        Filter nextFilter = filters.get(currentPosition++);
        nextFilter.doFilter(request, response, this);
```
else 문을 사용하지 않는 원칙 외에도 가시성을 위해서 종료시점에 return임을 명시해주는 것이 해당 코드의 가독성을 올리는데 도움을 줄 수 있을 것 같아요.